diff --git a/src/Makefile b/src/Makefile
index 55cc15847..6bc758884 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -380,6 +380,8 @@ else
     endef
 endif
 
+FINAL_CFLAGS+=-I../../../../Symlib/include -I../../kcut
+
 # Determine install/uninstall Redis symlinks for compatibility when
 # installing/uninstalling Valkey binaries (defaulting to `yes`)
 USE_REDIS_SYMLINKS?=yes
@@ -499,8 +501,8 @@ ifneq ($(strip $(PREV_FINAL_LDFLAGS)), $(strip $(FINAL_LDFLAGS)))
 endif
 
 # valkey-server
-$(SERVER_NAME): $(ENGINE_SERVER_OBJ)
-	$(SERVER_LD) -o $@ $^ ../deps/libvalkey/lib/libvalkey.a ../deps/lua/src/liblua.a ../deps/hdr_histogram/libhdrhistogram.a ../deps/fpconv/libfpconv.a $(FINAL_LIBS)
+$(SERVER_NAME): $(ENGINE_SERVER_OBJ) ../../kcut/libextension.a
+	$(SERVER_LD) -o $@ $^ ../deps/libvalkey/lib/libvalkey.a ../deps/lua/src/liblua.a ../deps/hdr_histogram/libhdrhistogram.a ../deps/fpconv/libfpconv.a $(FINAL_LIBS) -L ../../kcut -lc -lextension -lkallsyms -L ../../../../Symlib/dynam_build -lSym 
 
 # Valkey static library, used to compile against for unit testing
 $(ENGINE_LIB_NAME): $(ENGINE_SERVER_OBJ)
diff --git a/src/server.c b/src/server.c
index 838459336..faa721aef 100644
--- a/src/server.c
+++ b/src/server.c
@@ -79,6 +79,15 @@
 #include <sys/mman.h>
 #endif
 
+#include <dlfcn.h>
+#include <assert.h>
+#include <inttypes.h>
+
+#include "kernel.h"
+#include "evacuate.kh"
+#include "LINF/sym_all.h"
+#include "L1/stack_switch.h"
+
 #if defined(HAVE_SYSCTL_KIPC_SOMAXCONN) || defined(HAVE_SYSCTL_KERN_SOMAXCONN)
 #include <sys/sysctl.h>
 #endif
@@ -7135,12 +7144,49 @@ int iAmPrimary(void) {
             (server.cluster_enabled && clusterNodeIsPrimary(getMyClusterNode())));
 }
 
+static inline int
+resolve_sym(char *name, void **value)
+{
+  char *error;
+  dlerror();  // clear as per manpage
+  *value = dlsym(RTLD_DEFAULT, name);
+  error = dlerror();
+  if (error != NULL) {
+    fprintf(stderr, "%s\n", error);
+    return 0;
+  }
+  return 1;
+}
+
+
+static void evacuate()
+{
+  uintptr_t ktos;
+  int rc;
+  unsigned int cpu;
+
+  assert(getcpu(&cpu, NULL)==0);
+  
+  if (!resolve_sym("cpu_current_top_of_stack", (void **)&ktos)) {
+    printf("failed to resolve cpu_current_top_of_stack\n");
+    assert(0);
+  }
+
+  SYM_ON_KERN_STACK_DYNSYM_DO(ktos, 
+			      rc=acquire_exclusive_cpu(cpu,EVAC_KILL_NICELY));
+
+  printf("acquire_exclusive_cpu: %d\n", rc);
+  
+  assert(rc==0);
+}
+
+
 /* Main is marked as weak so that unit tests can use their own main function. */
 __attribute__((weak)) int main(int argc, char **argv) {
     struct timeval tv;
     int j;
     char config_from_stdin = 0;
-
+    evacuate(); // Evacuate core 
     /* We need to initialize our libraries, and the server configuration. */
 #ifdef INIT_SETPROCTITLE_REPLACEMENT
     spt_init(argc, argv);
diff --git a/src/socket.c b/src/socket.c
index 336b250cc..824bbba70 100644
--- a/src/socket.c
+++ b/src/socket.c
@@ -155,12 +155,30 @@ static void connSocketClose(connection *conn) {
     zfree(conn);
 }
 
+#include "../../kcut/kcut_tcpmsg.kh"
+#include "LINF/sym_all.h"
+extern void symbi_fast_lower(void);
+
 static int connSocketWrite(connection *conn, const void *data, size_t data_len) {
     /* Assert the main thread is not writing to a connection that is currently offloaded. */
     debugServerAssert(!(conn->flags & CONN_FLAG_ALLOW_ACCEPT_OFFLOAD) || !inMainThread() ||
                       ((client *)connGetPrivateData(conn))->io_write_state != CLIENT_PENDING_IO);
-
-    int ret = write(conn->fd, data, data_len);
+    struct iovec iov;
+    iov.iov_base = (void *)data;
+    iov.iov_len = data_len;
+
+    const int KCUT_THRESHOLD = 20;
+    static int kcut_cnt = 0;
+    int ret;    
+    if (kcut_cnt) {
+      sym_elevate();
+      ret = kcut_tcp_sendmsg(conn->fd, &iov);
+      symbi_fast_lower();
+      kcut_cnt--;
+    } else {
+      ret = write(conn->fd, data, data_len);
+      kcut_cnt = KCUT_THRESHOLD;
+    }
     if (ret < 0 && errno != EAGAIN) {
         conn->last_errno = errno;
 
@@ -191,9 +209,23 @@ static int connSocketRead(connection *conn, void *buf, size_t buf_len) {
     /* Assert the main thread is not reading from a connection that is currently offloaded. */
     debugServerAssert(!(conn->flags & CONN_FLAG_ALLOW_ACCEPT_OFFLOAD) || !inMainThread() ||
                       ((client *)connGetPrivateData(conn))->io_read_state != CLIENT_PENDING_IO);
-
-
-    int ret = read(conn->fd, buf, buf_len);
+    struct iovec iov;
+    iov.iov_base = (void *)buf;
+    iov.iov_len = buf_len;
+
+    const int KCUT_THRESHOLD = 20;
+    static int kcut_cnt = 0;
+    int ret;    
+    if (kcut_cnt) {
+      sym_elevate();
+      ret = kcut_tcp_recvmsg(conn->fd, &iov);
+      symbi_fast_lower();
+      kcut_cnt--;
+    } else {
+      ret = read(conn->fd, buf, buf_len);
+      kcut_cnt = KCUT_THRESHOLD;
+    }
+    
     if (!ret) {
         conn->state = CONN_STATE_CLOSED;
     } else if (ret < 0 && errno != EAGAIN) {
